library/src/main/java/com/hannesdorfmann/adapterdelegates3/AbsDiffDelegationAdapter.java

```diff
@@ -0,0 +1,109 @@
+package com.hannesdorfmann.adapterdelegates3;
+
+import android.support.annotation.NonNull;
+import android.support.v7.recyclerview.extensions.AsyncListDiffer;
+import android.support.v7.util.DiffUtil;
+import android.support.v7.widget.RecyclerView;
+import android.view.ViewGroup;
+
+import java.util.List;
+
+/**
+ * An implementation of an Adapter that already uses a {@link AdapterDelegatesManager} pretty same as
+ * {@link AbsDelegationAdapter} but also uses {@link AsyncListDiffer} from support library 27.0.1 for
+ * calculating diffs between old and new collections of items and does this on background thread.
+ * That means that now you should not carry about {@link RecyclerView.Adapter#notifyItemChanged(int)}
+ * and other methods of adapter, all you need to do is to submit a new list into adapter and all diffs will be
+ * calculated for you.
+ * You just have to add the {@link AdapterDelegate}s i.e. in the constructor of a subclass that inheritance from this
+ * class:
+ * <pre>
+ * {@code
+ *    class MyAdapter extends AbsDiffDelegationAdapter<MyDataSourceType> {
+ *        public MyAdapter() {
+ *            this.delegatesManager.add(new FooAdapterDelegate());
+ *            this.delegatesManager.add(new BarAdapterDelegate());
+ *        }
+ *    }
+ * }
+ * </pre>
+ *
+ * @param <T> The type of the item, must implement {@link DiffItem}
+ * @author Sergey Opivalov
+ */
+
+public abstract class AbsDiffDelegationAdapter<T extends DiffItem> extends RecyclerView.Adapter {
+
+    protected final AdapterDelegatesManager<List<T>> delegatesManager;
+    protected final AsyncListDiffer<T> differ;
+
+    public AbsDiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback) {
+        this.differ = new AsyncListDiffer<>(this, diffCallback);
+        this.delegatesManager = new AdapterDelegatesManager<>();
+    }
+
+    public AbsDiffDelegationAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback,
```

Comment:
If you can pass in any arbitarry DiffUtil.ItemCallback<T> then the restriction that it must be DiffItem is not useful anymore because one could write a DiffUtil.ItemCallback that is not using DiffItem.getItemId() nor DiffItem.getItemHash().

As I also pointed out in some other comments, I'm not sure if this library should contain something like DiffItem and a default DiffCallbackImpl respectively.

Reply:
But T is restricted with DiffItem interface in DiffDelegationAdapter class signature.